druid-shell/src/application.rs

```diff
@@ -36,44 +41,148 @@ pub trait AppHandler {
     fn command(&mut self, id: u32) {}
 }
 
-//TODO: we may want to make the user create an instance of this (Application::global()?)
-//but for now I'd like to keep changes minimal.
 /// The top level application object.
-pub struct Application(platform::Application);
+///
+/// This can be thought of as a reference and it can be safely cloned.
+#[derive(Clone)]
+pub struct Application {
+    state: Rc<RefCell<State>>,
+    pub(crate) platform_app: platform::Application,
+}
+
+/// Platform-independent `Application` state.
+struct State {
+    running: bool,
+}
+
+/// Used to ensure only one Application instance is ever created.
+static APPLICATION_CREATED: AtomicBool = AtomicBool::new(false);
+
+thread_local! {
+    /// A reference object to the current `Application`, if any.
+    static GLOBAL_APP: RefCell<Option<Application>> = RefCell::new(None);
+}
 
 impl Application {
-    pub fn new(handler: Option<Box<dyn AppHandler>>) -> Application {
-        Application(platform::Application::new(handler))
+    /// Create a new `Application`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if an `Application` has already been created.
+    ///
+    /// This may change in the future. See [druid#771] for discussion.
+    ///
+    /// [druid#771]: https://github.com/xi-editor/druid/issues/771
+    pub fn new() -> Application {
```

Comment:
I'm trying to think a bit about what the API should look like in this new world.

Some basic questions:

- should we require *explicit* initialization, or should we just lazily initialize in `global`?
- if we _do_ require explicit initialization, should that be a method like `new` that returns the application, or should it be a method like `init` that returns `Result<(), SomeError>`?, and then the actual application is always retrieved via `global`?
- Should `Application` be passed by value, or by reference? Having `global` return `&'static Application' (and having `Application: !Clone` perhaps better expresses the semantics of the program, where there is a single shared application).

Reply:
About your second bullet (In particular with gtk and perhaps macOS)

I think returning a `Result` might be good for cases where we are dealing with a remote application that was previously launched.  E.g. In the MacOS case there is `NSRunningApplication`, and in GTK the application is less functional, cannot display windows, but can send signals such as to open a new window to the remote application.

Being able to differentiate these via `Result<ApplicationInitializationStatus, SomeError>` or such might work for exposing this behavior on both platforms?

Reply:
I'll play around a bit with the `&'static Application` idea and see if and how well I can get it working.

Reply:
Okay I ran into a dead end with `&'static Application` but opened [a thread on zulip](https://xi.zulipchat.com/#narrow/stream/147926-druid/topic/druid.23763.20Application.20exit) for it.

Regarding initialization, I favor explicit because it gives clear control over the timing of when this significant work takes place, and if we plan on returning errors here in the future instead of panicking - then also a single place to handle those errors. The `global` function would always be fast and there wouldn't be a need to reason about whether one should check for initialization errors at every call location.

Naming wise, `new` vs `init`, I think `new` is better if we don't want to start painting ourselves into a strategy corner in terms of #771. That's because I think it makes more sense to have `new` if this function is going to be called more than once.

Return value wise I think changing it to `fn new() -> Result<Application, Error>` would be a good upgrade. This would allow the calling app to choose whether to panic or to do something else. Right now `druid-shell` is a bit too panic friendly in its initialization code.

Reply:
I pushed a new commit containing the `Result<Application, Error>` change.